File: ios/MangaReaderClone/Models/Manga.swift
// Data models for manga, chapters, and pages

import Foundation

enum MangaStatus: Int, Codable {
    case unknown = 0
    case ongoing = 1
    case completed = 2
    case licensed = 3
    case publishingFinished = 4
    case cancelled = 5
    case onHiatus = 6
}

struct Manga: Identifiable, Codable, Hashable {
    let id: String
    var title: String
    var coverUrl: String
    var description: String
    var author: String
    var artist: String
    var status: MangaStatus
    var genres: [String]
    var sourceId: String
    var url: String
    var inLibrary: Bool
    var lastUpdated: Date?

    init(
        id: String,
        title: String,
        coverUrl: String = "",
        description: String = "",
        author: String = "",
        artist: String = "",
        status: MangaStatus = .unknown,
        genres: [String] = [],
        sourceId: String = "",
        url: String = "",
        inLibrary: Bool = false,
        lastUpdated: Date? = nil
    ) {
        self.id = id
        self.title = title
        self.coverUrl = coverUrl
        self.description = description
        self.author = author
        self.artist = artist
        self.status = status
        self.genres = genres
        self.sourceId = sourceId
        self.url = url
        self.inLibrary = inLibrary
        self.lastUpdated = lastUpdated
    }

    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }

    static func == (lhs: Manga, rhs: Manga) -> Bool {
        lhs.id == rhs.id
    }
}

struct Chapter: Identifiable, Codable {
    let id: String
    let mangaId: String
    var name: String
    var url: String
    var chapterNumber: Double
    var dateUpload: Date?
    var scanlator: String
    var language: String
    var downloaded: Bool

    init(
        id: String,
        mangaId: String,
        name: String,
        url: String = "",
        chapterNumber: Double = 0,
        dateUpload: Date? = nil,
        scanlator: String = "",
        language: String = "en",
        downloaded: Bool = false
    ) {
        self.id = id
        self.mangaId = mangaId
        self.name = name
        self.url = url
        self.chapterNumber = chapterNumber
        self.dateUpload = dateUpload
        self.scanlator = scanlator
        self.language = language
        self.downloaded = downloaded
    }
}

struct Page: Identifiable {
    let id: Int  // index
    let imageUrl: String
    let chapterId: String
    var localPath: String?

    var isDownloaded: Bool { localPath != nil }

    init(index: Int, imageUrl: String, chapterId: String = "", localPath: String? = nil) {
        self.id = index
        self.imageUrl = imageUrl
        self.chapterId = chapterId
        self.localPath = localPath
    }
}

struct ExtensionInfo: Identifiable, Codable {
    var id: String { pkg }
    let pkg: String
    let name: String
    let apk: String
    let lang: String
    let code: Int
    let version: String
    let nsfw: Int
    let sources: [[String: AnyCodable]]?
    let iconUrl: String?

    enum CodingKeys: String, CodingKey {
        case pkg, name, apk, lang, code, version, nsfw, sources
        case iconUrl = "icon"
    }
}

/// Type-erased Codable wrapper for JSON values
struct AnyCodable: Codable {
    let value: Any

    init(_ value: Any) { self.value = value }

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if let int = try? container.decode(Int.self) { value = int }
        else if let double = try? container.decode(Double.self) { value = double }
        else if let string = try? container.decode(String.self) { value = string }
        else if let bool = try? container.decode(Bool.self) { value = bool }
        else { value = "" }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if let int = value as? Int { try container.encode(int) }
        else if let double = value as? Double { try container.encode(double) }
        else if let string = value as? String { try container.encode(string) }
        else if let bool = value as? Bool { try container.encode(bool) }
    }
}

struct ExtensionRepo {
    let url: String
    var name: String
    var extensions: [ExtensionInfo]
}

File: MangaReaderClone/models/manga.py
"""Data models for the manga reader application."""

from dataclasses import dataclass, field
from datetime import datetime
from enum import Enum
from typing import Optional


class MangaStatus(Enum):
    UNKNOWN = 0
    ONGOING = 1
    COMPLETED = 2
    LICENSED = 3
    PUBLISHING_FINISHED = 4
    CANCELLED = 5
    ON_HIATUS = 6


@dataclass
class Manga:
    id: str
    title: str
    cover_url: str = ""
    description: str = ""
    author: str = ""
    artist: str = ""
    status: MangaStatus = MangaStatus.UNKNOWN
    genres: list[str] = field(default_factory=list)
    source_id: str = ""
    url: str = ""
    in_library: bool = False
    last_updated: Optional[datetime] = None

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "title": self.title,
            "cover_url": self.cover_url,
            "description": self.description,
            "author": self.author,
            "artist": self.artist,
            "status": self.status.value,
            "genres": ",".join(self.genres),
            "source_id": self.source_id,
            "url": self.url,
            "in_library": self.in_library,
            "last_updated": self.last_updated.isoformat() if self.last_updated else None,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Manga":
        return cls(
            id=data["id"],
            title=data["title"],
            cover_url=data.get("cover_url", ""),
            description=data.get("description", ""),
            author=data.get("author", ""),
            artist=data.get("artist", ""),
            status=MangaStatus(data.get("status", 0)),
            genres=data.get("genres", "").split(",") if data.get("genres") else [],
            source_id=data.get("source_id", ""),
            url=data.get("url", ""),
            in_library=bool(data.get("in_library", False)),
            last_updated=datetime.fromisoformat(data["last_updated"]) if data.get("last_updated") else None,
        )


@dataclass
class Chapter:
    id: str
    manga_id: str
    name: str
    url: str
    chapter_number: float = 0.0
    date_upload: Optional[datetime] = None
    scanlator: str = ""
    language: str = "en"
    downloaded: bool = False

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "manga_id": self.manga_id,
            "name": self.name,
            "url": self.url,
            "chapter_number": self.chapter_number,
            "date_upload": self.date_upload.isoformat() if self.date_upload else None,
            "scanlator": self.scanlator,
            "language": self.language,
            "downloaded": self.downloaded,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Chapter":
        return cls(
            id=data["id"],
            manga_id=data["manga_id"],
            name=data["name"],
            url=data["url"],
            chapter_number=float(data.get("chapter_number", 0)),
            date_upload=datetime.fromisoformat(data["date_upload"]) if data.get("date_upload") else None,
            scanlator=data.get("scanlator", ""),
            language=data.get("language", "en"),
            downloaded=bool(data.get("downloaded", False)),
        )


@dataclass
class Page:
    index: int
    image_url: str
    chapter_id: str = ""
    local_path: str = ""

    @property
    def is_downloaded(self) -> bool:
        return bool(self.local_path)


@dataclass
class ExtensionInfo:
    pkg: str
    name: str
    apk: str
    lang: str
    code: int
    version: str
    nsfw: int = 0
    sources: list[dict] = field(default_factory=list)
    icon_url: str = ""

    @classmethod
    def from_dict(cls, data: dict) -> "ExtensionInfo":
        return cls(
            pkg=data.get("pkg", ""),
            name=data.get("name", ""),
            apk=data.get("apk", ""),
            lang=data.get("lang", "en"),
            code=data.get("code", 0),
            version=data.get("version", "1.0"),
            nsfw=data.get("nsfw", 0),
            sources=data.get("sources", []),
            icon_url=data.get("icon", ""),
        )


@dataclass
class ExtensionRepo:
    url: str
    name: str = ""
    extensions: list[ExtensionInfo] = field(default_factory=list)
